src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs

```diff
@@ -689,49 +863,202 @@ private static IEnumerable<double> LazySequence(double records, double start, do
+        internal static FormulaValue RoundFloat(IRContext irContext, NumberValue num, double doubleDigs, RoundType rt = RoundType.Default)
         {
-            var s = number < 0 ? -1d : 1d;
+            var number = num.Value;
+            var s = number < 0 ? -1d : 1d;
             var n = number * s;
-            var dg = digits < 0 ? (int)Math.Ceiling(digits) : (int)Math.Floor(digits);
 
-            if (dg < -15 || dg > 15 || number < -1e20d || number > 1e20d)
+            int dg = doubleDigs > int.MaxValue ? int.MaxValue :
+                          doubleDigs < int.MinValue ? int.MinValue :
+                                (int)doubleDigs;
+
+            if (dg < -15 || dg > 15 || number < -1e20d || number > 1e20d)
+            {
+                return num;
+            }
+
+            // Dividing by m, since multiplication was introducing floating point error
+            var m = Math.Pow(10d, dg);
+            var eps = 1 / (m * 1e12d); // used to manage rounding of 1.4499999999999999999996
+
+            switch (rt)
+            {
+                case RoundType.Default:
+                    return new NumberValue(irContext, s * Math.Floor((n + (1 / (2 * m)) + eps) * m) / m);
+                case RoundType.Down:
+                    return new NumberValue(irContext, s * Math.Floor(n * m) / m);
+                case RoundType.Up:
+                    return new NumberValue(irContext, s * Math.Ceiling(n * m) / m);
+            }
+
+            return CommonErrors.UnreachableCodeError(irContext);
+        }
+
+        private static readonly IReadOnlyList<decimal> DecPow10 = new decimal[]
+        {
+            1e+00m, 1e+01m, 1e+02m, 1e+03m, 1e+04m, 1e+05m, 1e+06m, 1e+07m, 1e+08m, 1e+09m,
+            1e+10m, 1e+11m, 1e+12m, 1e+13m, 1e+14m, 1e+15m, 1e+16m, 1e+17m, 1e+18m, 1e+19m,
+            1e+20m, 1e+21m, 1e+22m, 1e+23m, 1e+24m, 1e+25m, 1e+26m, 1e+27m, 1e+28m
+        };
+
+        private static readonly IReadOnlyList<decimal> DecNegPow10 = new decimal[]
+        {
+            1e-00m, 1e-01m, 1e-02m, 1e-03m, 1e-04m, 1e-05m, 1e-06m, 1e-07m, 1e-08m, 1e-09m,
+            1e-10m, 1e-11m, 1e-12m, 1e-13m, 1e-14m, 1e-15m, 1e-16m, 1e-17m, 1e-18m, 1e-19m,
+            1e-20m, 1e-21m, 1e-22m, 1e-23m, 1e-24m, 1e-25m, 1e-26m, 1e-27m, 1e-28m
+        };
+
+        // The algorithm for Decimal is different from that of Float because with less range we are going out of our way to avoid overflow
+        // At the time of this writing, the version of .NET being targeted only supports two varieties of MidpointRounding
+        // In the future, some of this can be replaced with built in support in decimal.Round()
+        internal static FormulaValue RoundDecimal(IRContext irContext, DecimalValue dec, double doubleDigs, RoundType roundType = RoundType.Default)
+        {
+            var signedNumber = dec.Value;
+            var sign = signedNumber < 0 ? -1m : 1m;

```

Comment:
>signedNumber < 0 ? -1m : 1m; [](http://example.com/codeflow?start=23&length=28)

decimal.Sign ?

Reply:
There is a Math.Sign but it returns an int.  I really don't want it to have to promote on the fly.  I think I'll leave it as.